tiledb/common/arithmetic.h

```diff
@@ -196,6 +196,20 @@ struct checked_arithmetic<uint64_t> {
       return -negated.value();
     }
   }
+
+  /**
+   * @return `a * b` if it can be represented as a `uint64_t` without undefined
+   * behavior, `std::nullopt` otherwise
+   */
+  static std::optional<uint64_t> mul(uint64_t a, uint64_t b) {
```

Comment:
You can perhaps use/modify `utils::math::safe_mul`.

Reply:
The scope of this is broader than it looks at first glance.  `safe_mul` handles floats, this new `mul` doesn't.  I do agree that it would be better to consolidate the functions; I am reluctant to do so in this already very large pull request.